akka-cluster-tools/src/test/scala/akka/cluster/singleton/ClusterSingletonLeavingSpeedSpec.scala

```diff
@@ -0,0 +1,137 @@
+      port = 0
+    }
+  }
+  """) {
+
+  // FIXME change number of systems to 3 if we keep this test, can be adjusted to higher when investigating
+  private val systems = Vector.fill[ActorSystem](40)(ActorSystem(system.name, system.settings.config))
+  private val probes = systems.map(TestProbe()(_))
+
+  override def expectedTestDuration: FiniteDuration = 10.minutes
+
+  def join(from: ActorSystem, to: ActorSystem, probe: ActorRef): Unit = {
+    from.actorOf(
+      ClusterSingletonManager.props(
+        singletonProps = TheSingleton.props(probe),
+        terminationMessage = PoisonPill,
+        settings = ClusterSingletonManagerSettings(from)),
+      name = "echo")
+
+    Cluster(from).join(Cluster(to).selfAddress)
+    within(15.seconds) {
+      awaitAssert {
+        Cluster(from).state.members.map(_.uniqueAddress) should contain(Cluster(from).selfUniqueAddress)
+        Cluster(from).state.members.map(_.status) should ===(Set(MemberStatus.Up))
+      }
+    }
+  }
+
+  "ClusterSingleton that is leaving" must {
+    "join cluster" in {
+      systems.indices.foreach { i ⇒
+        join(systems(i), systems.head, probes(i).ref)
+      }
+      // leader is most likely on system, lowest port
+      join(system, systems.head, testActor)
+
+      probes(0).expectMsg("started")
+    }
+
+    "quickly hand-over to next oldest" in {
+
+      val durations = systems.indices.map { i ⇒
+        val t0 = System.nanoTime()
+        val leaveAddress = Cluster(systems(i)).selfAddress
+        //        Cluster(systems(i)).leave(leaveAddress)
+        //        CoordinatedShutdown(systems(i)).run(CoordinatedShutdown.ClusterLeavingReason)
+        Cluster(system).leave(leaveAddress)
+        probes(i).expectMsg(10.seconds, "stopped")
+        val stoppedDuration = (System.nanoTime() - t0).nanos
+        val startedProbe = if (i == systems.size - 1) this else probes(i + 1)
+        startedProbe.expectMsg(10.seconds, "started")
+        val startedDuration = (System.nanoTime() - t0).nanos
+
+        within(10.seconds) {
+          awaitAssert {
+            Cluster(systems(i)).isTerminated should ===(true)
+            Cluster(system).state.members.map(_.address) should not contain leaveAddress
+            systems.foreach { sys ⇒
+              if (!Cluster(sys).isTerminated)
+                Cluster(sys).state.members.map(_.address) should not contain leaveAddress
+            }
+          }
+        }
+
+        println(s"Singleton $i stopped in ${stoppedDuration.toMillis} ms, started in ${startedDuration.toMillis} ms, " +
+          s"diff ${(startedDuration - stoppedDuration).toMillis} ms")
+
+        (stoppedDuration, startedDuration)
+      }
+
+      durations.zipWithIndex.foreach {
+        case ((stoppedDuration, startedDuration), i) ⇒
+          println(s"Singleton $i stopped in ${stoppedDuration.toMillis} ms, started in ${startedDuration.toMillis} ms, " +
+            s"diff ${(startedDuration - stoppedDuration).toMillis} ms")
+      }
```

Comment:
@412b I tried to create a test simulating the slow hand over that you reported. Unfortunately I have not been able to see any problems. My results of this is: 
```
Singleton 0 stopped in 6062 ms, started in 6063 ms, diff 1 ms
Singleton 1 stopped in 5468 ms, started in 5468 ms, diff 0 ms
Singleton 2 stopped in 7086 ms, started in 7087 ms, diff 0 ms
Singleton 3 stopped in 6084 ms, started in 6084 ms, diff 0 ms
Singleton 4 stopped in 6367 ms, started in 6367 ms, diff 0 ms
Singleton 5 stopped in 5279 ms, started in 5279 ms, diff 0 ms
Singleton 6 stopped in 5514 ms, started in 5515 ms, diff 0 ms
Singleton 7 stopped in 6383 ms, started in 6383 ms, diff 0 ms
Singleton 8 stopped in 4364 ms, started in 4365 ms, diff 0 ms
Singleton 9 stopped in 4937 ms, started in 4938 ms, diff 0 ms
Singleton 10 stopped in 4463 ms, started in 4463 ms, diff 0 ms
Singleton 11 stopped in 5033 ms, started in 5034 ms, diff 0 ms
Singleton 12 stopped in 5465 ms, started in 5466 ms, diff 0 ms
Singleton 13 stopped in 4716 ms, started in 4717 ms, diff 0 ms
Singleton 14 stopped in 5868 ms, started in 5869 ms, diff 0 ms
Singleton 15 stopped in 5116 ms, started in 5116 ms, diff 0 ms
Singleton 16 stopped in 4651 ms, started in 4652 ms, diff 0 ms
Singleton 17 stopped in 7847 ms, started in 7848 ms, diff 0 ms
Singleton 18 stopped in 6942 ms, started in 6943 ms, diff 0 ms
Singleton 19 stopped in 5517 ms, started in 5518 ms, diff 0 ms
Singleton 20 stopped in 5797 ms, started in 5798 ms, diff 0 ms
Singleton 21 stopped in 4500 ms, started in 4500 ms, diff 0 ms
Singleton 22 stopped in 3823 ms, started in 3823 ms, diff 0 ms
Singleton 23 stopped in 3854 ms, started in 3854 ms, diff 0 ms
Singleton 24 stopped in 4042 ms, started in 4043 ms, diff 0 ms
Singleton 25 stopped in 2214 ms, started in 2214 ms, diff 0 ms
Singleton 26 stopped in 4852 ms, started in 4853 ms, diff 0 ms
Singleton 27 stopped in 5138 ms, started in 5138 ms, diff 0 ms
Singleton 28 stopped in 3657 ms, started in 3658 ms, diff 0 ms
Singleton 29 stopped in 3836 ms, started in 3837 ms, diff 0 ms
Singleton 30 stopped in 4919 ms, started in 4920 ms, diff 0 ms
Singleton 31 stopped in 3146 ms, started in 3146 ms, diff 0 ms
Singleton 32 stopped in 4884 ms, started in 4884 ms, diff 0 ms
Singleton 33 stopped in 3607 ms, started in 3608 ms, diff 0 ms
Singleton 34 stopped in 5903 ms, started in 5904 ms, diff 0 ms
Singleton 35 stopped in 3979 ms, started in 3979 ms, diff 0 ms
Singleton 36 stopped in 3009 ms, started in 3010 ms, diff 0 ms
Singleton 37 stopped in 2906 ms, started in 2907 ms, diff 0 ms
Singleton 38 stopped in 1929 ms, started in 1929 ms, diff 0 ms
Singleton 39 stopped in 900 ms, started in 900 ms, diff 0 ms
```

Note that the old singleton instance is still alive during the leaving-exiting and that's why the diff is so small. It's stopped when they can exchange the hand over messages. Therefore I wonder what you measure? Why does it matter that it takes a few seconds before the hand over?

I have an optimization in mind, but wanted to understand current performance first. We could send gossip to the 2 oldest nodes (also per role) when the leader changes the leaving node to Exiting.

Reply:
Pushed an optimization, which results in:
```
Singleton 0 stopped in 3817 ms, started in 3821 ms, diff 4 ms
Singleton 1 stopped in 3823 ms, started in 3827 ms, diff 3 ms
Singleton 2 stopped in 3567 ms, started in 3571 ms, diff 3 ms
Singleton 3 stopped in 3165 ms, started in 3167 ms, diff 2 ms
Singleton 4 stopped in 4163 ms, started in 4163 ms, diff 0 ms
Singleton 5 stopped in 3619 ms, started in 3620 ms, diff 0 ms
Singleton 6 stopped in 5997 ms, started in 6000 ms, diff 3 ms
Singleton 7 stopped in 4301 ms, started in 4304 ms, diff 2 ms
Singleton 8 stopped in 3128 ms, started in 3130 ms, diff 2 ms
Singleton 9 stopped in 3320 ms, started in 3323 ms, diff 2 ms
Singleton 10 stopped in 4495 ms, started in 4498 ms, diff 2 ms
Singleton 11 stopped in 3532 ms, started in 3535 ms, diff 2 ms
Singleton 12 stopped in 3561 ms, started in 3563 ms, diff 2 ms
Singleton 13 stopped in 4437 ms, started in 4440 ms, diff 2 ms
Singleton 14 stopped in 2661 ms, started in 2665 ms, diff 3 ms
Singleton 15 stopped in 3679 ms, started in 3682 ms, diff 2 ms
Singleton 16 stopped in 3783 ms, started in 3786 ms, diff 2 ms
Singleton 17 stopped in 2798 ms, started in 2800 ms, diff 2 ms
Singleton 18 stopped in 3607 ms, started in 3613 ms, diff 5 ms
Singleton 19 stopped in 2826 ms, started in 2827 ms, diff 0 ms
Singleton 20 stopped in 4914 ms, started in 4915 ms, diff 0 ms
Singleton 21 stopped in 4793 ms, started in 4794 ms, diff 0 ms
Singleton 22 stopped in 3911 ms, started in 3912 ms, diff 0 ms
Singleton 23 stopped in 2629 ms, started in 2629 ms, diff 0 ms
Singleton 24 stopped in 2048 ms, started in 2049 ms, diff 0 ms
Singleton 25 stopped in 3668 ms, started in 3668 ms, diff 0 ms
Singleton 26 stopped in 4569 ms, started in 4570 ms, diff 0 ms
Singleton 27 stopped in 2354 ms, started in 2354 ms, diff 0 ms
Singleton 28 stopped in 3664 ms, started in 3664 ms, diff 0 ms
Singleton 29 stopped in 3664 ms, started in 3665 ms, diff 0 ms
Singleton 30 stopped in 3466 ms, started in 3466 ms, diff 0 ms
Singleton 31 stopped in 2577 ms, started in 2577 ms, diff 0 ms
Singleton 32 stopped in 2557 ms, started in 2558 ms, diff 0 ms
Singleton 33 stopped in 2425 ms, started in 2425 ms, diff 0 ms
Singleton 34 stopped in 3487 ms, started in 3487 ms, diff 0 ms
Singleton 35 stopped in 2477 ms, started in 2477 ms, diff 0 ms
Singleton 36 stopped in 2436 ms, started in 2436 ms, diff 0 ms
Singleton 37 stopped in 506 ms, started in 507 ms, diff 0 ms
Singleton 38 stopped in 1448 ms, started in 1448 ms, diff 0 ms
Singleton 39 stopped in 447 ms, started in 448 ms, diff 0 ms
```